File: Java/src/generics/abs/AbsVertex.java
package generics.abs;

import java.util.HashSet;

public abstract class AbsVertex<G extends AbsGraph<G, V, E>, V extends AbsVertex<G, V, E>, E extends AbsEdge<G, V, E>>
{
    private Integer label;
    private HashSet<E> incidentEdges;
    private G graph;

    public AbsVertex(Integer label)
    {
        this.label = label;
        incidentEdges = new HashSet<E>();
    }
    
    public void unbindGraph()
    {
        for (E e : incidentEdges)
            graph.removeEdge(e);

        this.graph = null;
    }

    public G getGraph()
    {
        return graph;
    }
            
    public void setGraph(G g)
    {
        if (graph != null) unbindGraph();
        this.graph = g;
    }

    public HashSet<E> getIncidentEdges()
    {
        return incidentEdges;
    }

    public String toString()
    {
        return Integer.toString(label);
    }
}

File: scala/src/generics/abs/AbsVertex.scala
package generics.abs

import scala.collection.mutable.HashSet

class AbsVertex[G <: AbsGraph[G, V, E], V <: AbsVertex[G, V, E], E <: AbsEdge[G, V, E]](private var label:Integer) {

  def unbindGraph(): Unit = {
    incidentEdges.foreach(e => graph.removeEdge(e));

    this.graph = null.asInstanceOf[G]
  }

  def getGraph: G = graph

  def setGraph(g: G): Unit = {
    if (graph != null)
      unbindGraph()

    this.graph = g
  }

  def getIncidentEdges: HashSet[E] =  incidentEdges

  override def toString: String = label.toString()
 
  private var incidentEdges: HashSet[E] = new HashSet[E]
  private var graph: G = _
}
